File: testbed_ruby/test/associations/optional_one_to_optional_n_test.rb
require 'test/unit'
require './src-gen-umple/student_y'
require './src-gen-umple/mentor_y'
require './src-gen-umple/program_y'

module CruiseTest
class OptionalOneToOptionalNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentY.new(99)
    assert_equal(nil,s.get_mentor)
  end
  
  def test_add_student
    m = MentorY.new("blah")
    s = StudentY.new(99)
    m.add_student(s)
    
    assert_equal(m,s.get_mentor)
    assert_equal(s,m.get_student(0))
  end
  
  def test_add_studentCannotAddForever
    m = MentorY.new("blah")
    assert_equal(true,m.add_student(StudentY.new(99)))
    assert_equal(true,m.add_student(StudentY.new(98)))
    assert_equal(true,m.add_student(StudentY.new(97)))
    assert_equal(true,m.add_student(StudentY.new(96)))
    assert_equal(false,m.add_student(StudentY.new(95)))
  end
  
  def test_replaceMentor
    m = MentorY.new("blah")
    m2 = MentorY.new("blah2")
  
    s = StudentY.new(99)
    m.add_student(s)
    s.set_mentor(m2)
    
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)
  end 
  
  def test_cannotReplaceIfFull
    m = MentorY.new("blah")
    m2 = MentorY.new("blah2")
  
    m2.add_student(StudentY.new(99))
    m2.add_student(StudentY.new(98))
    m2.add_student(StudentY.new(97))
    m2.add_student(StudentY.new(96))
    
    s = StudentY.new(95)
    
    assert_equal(true,m.add_student(s))
    assert_equal(false,s.set_mentor(m2))
    
    assert_equal(m,s.get_mentor)
    assert_equal(4,m2.number_of_students)
    assert_equal(1,m.number_of_students)
  end 
  
  def test_addToNewMentor
    m = MentorY.new("blah")
    m2 = MentorY.new("blah2")
    s = StudentY.new(99)
    
    assert_equal(true, m.add_student(s))
    assert_equal(m,s.get_mentor)

    assert_equal(true, m2.add_student(s))
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)
  end  
  
  def test_cannotAddToFullMentor
    m = MentorY.new("blah")
    m2 = MentorY.new("blah2")
    s = StudentY.new(99)
    
    m.add_student(StudentY.new(98))
    m.add_student(StudentY.new(97))
    m.add_student(StudentY.new(96))
    m.add_student(StudentY.new(95))
    
    m2.add_student(s)
    assert_equal(false,m.add_student(s))
    assert_equal(m2,s.get_mentor)
    assert_equal(4,m.number_of_students)
    assert_equal(1,m2.number_of_students)
  end
  
  def test_removeFromExistingMentor
    m = MentorY.new("blah")
    m2 = MentorY.new("blah2")
  
    s1 = StudentY.new(99)
    s2 = StudentY.new(98)
    s3 = StudentY.new(97)
    
    m.add_student(s1)
    m.add_student(s2)
    m.add_student(s3)
  
    assert_equal(3,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    
    m2.add_student(s1)
  
    assert_equal(2,m.number_of_students)
    assert_equal(s2,m.get_student(0))
    assert_equal(s3,m.get_student(1))
    
    assert_equal(1,m2.number_of_students)
    assert_equal(s1,m2.get_student(0))
    
    assert_equal(m2,s1.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
  end   
  
  def test_remove_student
    m = MentorY.new("blah")
    s = StudentY.new(99)
    
    m.add_student(s)
    m.remove_student(s)
    
    assert_equal(nil,s.get_mentor)
    assert_equal(0,m.number_of_students)
    
  end  
  
  def test_maximum_number_of_students
    assert_equal(4,MentorY::maximum_number_of_students)
  end
  
end
end

File: testbed_php/test/associations/OptionalOneToOptionalNTest.php
<?php

class OptionalOneToOptionalNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentY(99);
    $this->assertEqual(null,$s->getMentor());
  }
  

  public function test_addStudent()
  {
    $m = new MentorY("blah");
    $s = new StudentY(99);
    $m->addStudent($s);
    
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($s,$m->getStudent_index(0));
  }


  public function test_addStudentCannotAddForever()
  {
    $m = new MentorY("blah");
    $this->assertEqual(true,$m->addStudent(new StudentY(99)));
    $this->assertEqual(true,$m->addStudent(new StudentY(98)));
    $this->assertEqual(true,$m->addStudent(new StudentY(97)));
    $this->assertEqual(true,$m->addStudent(new StudentY(96)));
    $this->assertEqual(false,$m->addStudent(new StudentY(95)));
  }

  

  public function test_replaceMentor()
  {
    $m = new MentorY("blah");
    $m2 = new MentorY("blah2");

    $s = new StudentY(99);
    $m->addStudent($s);
    $s->setMentor($m2);
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent_index(0));
    $this->assertEqual(0,$m->numberOfStudents());
  } 


  public function test_cannotReplaceIfFull()
  {
    $m = new MentorY("blah");
    $m2 = new MentorY("blah2");

    $m2->addStudent(new StudentY(99));
    $m2->addStudent(new StudentY(98));
    $m2->addStudent(new StudentY(97));
    $m2->addStudent(new StudentY(96));
    
    $s = new StudentY(95);
    
    $this->assertEqual(true,$m->addStudent($s));
    $this->assertEqual(false,$s->setMentor($m2));
    
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(4,$m2->numberOfStudents());
    $this->assertEqual(1,$m->numberOfStudents());
  } 
  

  public function test_addToNewMentor()
  {
    $m = new MentorY("blah");
    $m2 = new MentorY("blah2");
    $s = new StudentY(99);
    
    $m->addStudent($s);
    $m2->addStudent($s);
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent_index(0));
    $this->assertEqual(0,$m->numberOfStudents());
  }  
  

  public function test_cannotAddToFullMentor()
  {
    $m = new MentorY("blah");
    $m2 = new MentorY("blah2");
    $s = new StudentY(99);
    
    $m->addStudent(new StudentY(98));
    $m->addStudent(new StudentY(97));
    $m->addStudent(new StudentY(96));
    $m->addStudent(new StudentY(95));
    
    $m2->addStudent($s);
    $this->assertEqual(false,$m->addStudent($s));
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual(4,$m->numberOfStudents());
    $this->assertEqual(1,$m2->numberOfStudents());
  }
  

  public function test_removeFromExistingMentor()
  {
    $m = new MentorY("blah");
    $m2 = new MentorY("blah2");

    $s1 = new StudentY(99);
    $s2 = new StudentY(98);
    $s3 = new StudentY(97);
    
    $m->addStudent($s1);
    $m->addStudent($s2);
    $m->addStudent($s3);

    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    
    $m2->addStudent($s1);

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($s2,$m->getStudent_index(0));
    $this->assertEqual($s3,$m->getStudent_index(1));
    
    $this->assertEqual(1,$m2->numberOfStudents());
    $this->assertEqual($s1,$m2->getStudent_index(0));
    
    $this->assertEqual($m2,$s1->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
  }   
  

  public function test_removeStudent()
  {
    $m = new MentorY("blah");
    $s = new StudentY(99);
    
    $m->addStudent($s);
    $m->removeStudent($s);
    
    $this->assertEqual(null,$s->getMentor());
    $this->assertEqual(0,$m->numberOfStudents());
    
  }  


  public function test_maximumNumberOfStudents()
  {
    $this->assertEqual(4,MentorY::maximumNumberOfStudents());
  }
  
  
}
